File: biddo_web/src/core/repositories/auction.ts
import { Auction, LatLng } from '../domain/auction'
import { RequestMaker, RequestType } from '../services/request-maker'
import { Location } from '../domain/location'
import { GENERAL_ERRORS } from '../../constants/errors'

export enum AuctionsSortBy {
  newest,
  oldest,
  priceAsc,
  priceDesc,
}

export interface CountAuctionsFilterParams {
  categories?: string[]
  subCategories?: string[]
  locationIds?: string[]
  activeOnly?: boolean
  includeMyAuctions?: boolean
  minPrice?: number | string
  maxPrice?: number | string
  usedCurrencyId?: string
}

export interface FilterAuctionsParams {
  page: number
  categories?: string[]
  subCategories?: string[]
  locationIds?: string[]
  activeOnly?: boolean
  pageSize?: number
  query?: string
  sortBy?: AuctionsSortBy
  includeMyAuctions?: boolean
  minPrice?: number | string
  maxPrice?: number | string
  promotedOnly?: boolean
  usedCurrencyId?: string
  started?: boolean
}

class AuctionRepository {
  async loadSummary(auctionId: string): Promise<Auction | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: `/auction/summary/${auctionId}`,
      })) as Record<string, unknown>
      return Auction.fromJSON(response)
    } catch (error) {
      console.error('Error getting auction summary:', error)
      return null
    }
  }

  async loadDetails(auctionId: string): Promise<Auction | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: `/auction/details/${auctionId}`,
      })) as Record<string, unknown>
      return Auction.fromJSON(response)
    } catch (error) {
      console.error('Error getting auction details:', error)
      return null
    }
  }

  async storeLastSeenAuction(auctionId: string): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: `/lastSeen`,
        payload: JSON.stringify({ auctionId }),
        contentType: 'application/json',
      })
      return true
    } catch (error) {
      console.error('Error storing last seen auction:', error)
      return false
    }
  }

  async translateDetails(
    auctionId: string,
    language: string
  ): Promise<{ title: string; description: string } | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: `/auction/translate/${auctionId}/${language}`,
      })) as Record<string, unknown>
      return response as { title: string; description: string }
    } catch (error) {
      console.error('Error translating auction details:', error)
      return null
    }
  }

  async loadAuctionLocations(): Promise<Location[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: '/location/all',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Location.fromJSON(el))
    } catch (error) {
      console.error('Error getting locations:', error)
      return []
    }
  }

  async remove(auctionId: string): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.DELETE,
        path: `/auction/${auctionId}`,
      })
      return true
    } catch (error) {
      console.error('Error removing auction:', error)
      return false
    }
  }

  async loadRecommendations(page = 0, perPage = 10): Promise<Auction[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: '/auction-similarities',
        payload: JSON.stringify({ page, perPage }),
        contentType: 'application/json',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error getting recommendations:', error)
      return []
    }
  }

  async loadByProximity(latlng: LatLng, mainCategoryId: string, distance = 10): Promise<Auction[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: `/auction/byLocationProximity/${latlng.lat}/${latlng.lng}/${mainCategoryId}/${distance}`,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error getting auctions by proximity:', error)
      return []
    }
  }

  async promote(auctionId: string): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.PUT,
        path: `/auction/promote/${auctionId}`,
      })
      return true
    } catch (error) {
      console.error('Error promoting auction:', error)
      return false
    }
  }

  async countForAccountByBidStatus(status: string, query = ''): Promise<number> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `/auction/byBid/${status}/count`,
        method: RequestType.POST,
        payload: JSON.stringify({ query }),
      })) as { count: number }
      return response.count
    } catch (error) {
      console.error('Error counting auctions for account:', error)
      return 0
    }
  }

  async countForAccount(status: string, query = ''): Promise<number> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `/auction/all/account/${status}/count`,
        method: RequestType.POST,
        payload: JSON.stringify({ query }),
      })) as { count: number }
      return response.count
    } catch (error) {
      console.error('Error counting auctions for account:', error)
      return 0
    }
  }

  async loadForAccountByBidStatus(
    status: string,
    page = 0,
    pageSize = 10,
    query = '',
    sortBy: AuctionsSortBy = AuctionsSortBy.oldest
  ) {
    const { orderBy, orderDirection } = this.getOrderParams(sortBy)

    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: `/auction/byBid/${status}`,
        payload: JSON.stringify({
          page,
          perPage: pageSize,
          query,
          orderBy,
          orderDirection,
        }),
        contentType: 'application/json',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error getting auctions for account:', error)
      return []
    }
  }

  async loadForAccount(
    status: string,
    page = 0,
    pageSize = 10,
    query = '',
    sortBy: AuctionsSortBy = AuctionsSortBy.oldest
  ): Promise<Auction[]> {
    const { orderBy, orderDirection } = this.getOrderParams(sortBy)

    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: `/auction/all/account/${status}`,
        payload: JSON.stringify({
          page,
          perPage: pageSize,
          query,
          orderBy,
          orderDirection,
        }),
        contentType: 'application/json',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error getting auctions for account:', error)
      return []
    }
  }

  async search(query: string, page = 0, perPage = 10): Promise<Auction[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: `/auction/search/${query}/${page}/${perPage}`,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error searching auctions:', error)
      return []
    }
  }

  async loadLatestAuctions(): Promise<Auction[] | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.GET,
        path: '/auction/latest',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error getting latest auctions:', error)
      return null
    }
  }

  async countFilter(params: CountAuctionsFilterParams): Promise<number> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: '/auction/filter/count',
        payload: JSON.stringify({
          ...params,
        }),
        contentType: 'application/json',
      })) as { count: number }
      return Number(response.count)
    } catch (error) {
      console.error('Error counting filtered auctions:', error)
      return 0
    }
  }

  async countActiveForAccount(accountId: string, query = ''): Promise<number> {
    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: `/auction/byAccount/active/count/${accountId}`,
        payload: JSON.stringify({ query }),
      })) as { count: number }
      return response.count
    } catch (error) {
      console.error('Error counting active auctions for account:', error)
      return 0
    }
  }

  async loadActiveForAccount(
    accountId: string,
    page = 0,
    pageSize = 10,
    query = '',
    sortBy: AuctionsSortBy = AuctionsSortBy.newest
  ): Promise<Auction[]> {
    const { orderBy, orderDirection } = this.getOrderParams(sortBy)

    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: `/auction/byAccount/active/${accountId}`,
        payload: JSON.stringify({
          page,
          perPage: pageSize,
          query,
          orderBy,
          orderDirection,
        }),
        contentType: 'application/json',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error getting active auctions for account:', error)
      return []
    }
  }

  async loadFilteredAuctions(params: FilterAuctionsParams): Promise<Auction[]> {
    const { sortBy = AuctionsSortBy.oldest, pageSize = 20 } = params
    const { orderBy, orderDirection } = this.getOrderParams(sortBy)

    try {
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: '/auction/filter/auctions',
        payload: JSON.stringify({
          ...params,
          orderBy,
          orderDirection,
          perPage: pageSize,
        }),
        contentType: 'application/json',
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Auction.fromJSON(el))
    } catch (error) {
      console.error('Error loading filtered auctions:', error)
      return []
    }
  }

  async update(params: Record<string, unknown>): Promise<Auction | null> {
    try {
      const data = await this.generateFormDataForCreateOrUpdate(params)
      const response = (await RequestMaker.makeRequest({
        method: RequestType.PUT,
        path: `/auction/${params.id}`,
        payload: data,
        contentType: 'multipart/form-data',
      })) as Record<string, unknown>
      return Auction.fromJSON(response)
    } catch (error) {
      this.handleError(error as Error)
      return null
    }
  }

  async create(params: Record<string, unknown>): Promise<Auction | null> {
    try {
      const data = await this.generateFormDataForCreateOrUpdate(params)
      const response = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: '/auction',
        payload: data,
        contentType: 'multipart/form-data',
      })) as Record<string, unknown>
      return Auction.fromJSON(response)
    } catch (error) {
      this.handleError(error as Error)
      return null
    }
  }

  private async generateFormDataForCreateOrUpdate(
    params: Record<string, unknown>
  ): Promise<FormData> {
    const form = new FormData()
    Object.keys(params).forEach((key) => {
      if (key === 'id') {
        return
      }
      if (key === 'files') {
        const files = params[key] as File[]
        files.forEach((file) => {
          form.append('files', file)
        })
      } else {
        form.set(
          key,
          typeof params[key] === 'string' ? (params[key] as string) : JSON.stringify(params[key])
        )
      }
    })

    return form
  }

  private handleError(error: Error) {
    const response = JSON.stringify(error)

    if (response.includes('TOO_MANY')) {
      return { error: true, message: GENERAL_ERRORS.TOO_MANY_ASSETS }
    }
    if (response.includes('ASSET_TOO_BIG')) {
      return { error: true, message: GENERAL_ERRORS.ASSET_TOO_BIG }
    }
    if (response.includes('ASSET_NOT_SUPPORTED')) {
      return { error: true, message: GENERAL_ERRORS.ASSET_NOT_SUPPORTED }
    }

    return { error: true, message: 'Could not create or update auction' }
  }

  private getOrderParams(sortBy: AuctionsSortBy): {
    orderBy: string
    orderDirection: string
  } {
    switch (sortBy) {
      case AuctionsSortBy.newest:
        return { orderBy: 'createdAt', orderDirection: 'desc' }
      case AuctionsSortBy.oldest:
        return { orderBy: 'createdAt', orderDirection: 'asc' }
      case AuctionsSortBy.priceDesc:
        return { orderBy: 'lastPrice', orderDirection: 'desc' }
      case AuctionsSortBy.priceAsc:
        return { orderBy: 'lastPrice', orderDirection: 'asc' }
      default:
        return { orderBy: 'createdAt', orderDirection: 'desc' }
    }
  }
}

const auctionRepositoryInstance = new AuctionRepository()
export { auctionRepositoryInstance as AuctionRepository }

File: biddo_client/lib/core/repositories/auction.dart
import 'dart:convert';

import 'package:biddo/core/repositories/base.dart';
import 'package:dio/dio.dart';
import 'package:easy_localization/easy_localization.dart';
import 'package:get/get.dart';
import 'package:dio/dio.dart' as dio_lib;
import 'package:google_maps_flutter/google_maps_flutter.dart';

import '../controllers/flash.dart';
import '../models/auction.dart';
import '../models/auction_map_data.dart';
import '../models/location.dart';
import '../services/image_compress.dart';

enum AuctionsSortBy {
  newest,
  oldest,
  priceAsc,
  priceDesc,
}

class AuctionRepository {
  final dio = Get.find<Api>();
  final imageCompressService = Get.find<ImageCompressService>();
  final flashController = Get.find<FlashController>();

  Future<Auction?> loadDetails(String auctionId) async {
    try {
      var response = await dio.api.get('/auction/details/$auctionId');
      return Auction.fromJSON(response.data);
    } catch (error, stackTrace) {
      print('Error getting auction details: $error $stackTrace');
      return null;
    }
  }

  Future<List<Auction>?> loadSimilarAuctions(String auctionId,
      [int page = 0, int perPage = 10]) async {
    try {
      var response = await dio.api.post('/auction-similarities/similar', data: {
        'auctionId': auctionId,
        "page": page,
        "perPage": perPage,
      });
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Error getting similar auctions: $error, $stackTrace');
      return [];
    }
  }

  Future<AuctionTranslationResult?> translateDetails(
      String auctionId, String lang) async {
    try {
      var response = await dio.api.get('/auction/translate/$auctionId/$lang');
      return AuctionTranslationResult.fromJSON(response.data);
    } catch (error, stackTrace) {
      print('Error translating auction details: $error $stackTrace');
      return null;
    }
  }

  Future<List<Location>> loadAuctionLocations() async {
    try {
      var response = await dio.api.get('/location/all');
      return List<Location>.from(
        response.data.map(
          (el) => Location.fromJSON(el),
        ),
      );
    } catch (error) {
      print('Error getting locations: $error');
      return [];
    }
  }

  Future<bool> delete(String auctionId) async {
    try {
      await dio.api.delete('/auction/$auctionId');
      return true;
    } catch (error) {
      print('Error removing auction: $error');
      return false;
    }
  }

  Future<List<Auction>> loadRecommendations([
    int page = 0,
    int perPage = 10,
  ]) async {
    try {
      var response = await dio.api.post('/auction-similarities', data: {
        "page": page,
        "perPage": perPage,
      });
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Error getting recommendations: $error, $stackTrace');
      return [];
    }
  }

  Future<List<Auction>> loadByProximity(LatLng latlng, String mainCategoryId,
      [int distance = 10]) async {
    try {
      var response = await dio.api.get(
        '/auction/byLocationProximity/${latlng.latitude}/${latlng.longitude}/$mainCategoryId/$distance',
      );

      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error) {
      print('Error getting auctions by proximity: $error');
      return [];
    }
  }

  Future<bool> promote(String auctionId) async {
    try {
      await dio.api.put('/auction/promote/$auctionId');
      return true;
    } catch (error) {
      print('Error promoting auction: $error');
      return false;
    }
  }

  Future<List<Auction>> loadForAccount(
    String status, [
    int page = 0,
    int pageSize = 10,
    String query = '',
    AuctionsSortBy? sortBy = AuctionsSortBy.oldest,
  ]) async {
    try {
      String orderBy;
      String orderDirection;

      switch (sortBy) {
        case AuctionsSortBy.newest:
          orderBy = 'createdAt';
          orderDirection = 'desc';
        case AuctionsSortBy.oldest:
          orderBy = 'createdAt';
          orderDirection = 'asc';
        case AuctionsSortBy.priceDesc:
          orderBy = 'lastPrice';
          orderDirection = 'desc';
        case AuctionsSortBy.priceAsc:
          orderBy = 'lastPrice';
          orderDirection = 'asc';
        case null:
          orderBy = 'createdAt';
          orderDirection = 'desc';
      }

      if (status.contains('bid')) {
        var response = await dio.api
            .post('/auction/byBid/${status.replaceAll('bid-', '')}', data: {
          "page": page,
          "perPage": pageSize,
          "query": query,
          "orderBy": orderBy,
          "orderDirection": orderDirection,
        });
        return List<Auction>.from(
          response.data.map(
            (el) => Auction.fromJSON(el),
          ),
        );
      }

      var response = await dio.api.post('/auction/all/account/$status', data: {
        "page": page,
        "perPage": pageSize,
        "query": query,
        "orderBy": orderBy,
        "orderDirection": orderDirection,
      });
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Error getting auctions for account: $error, $stackTrace');
      return [];
    }
  }

  Future<List<Auction>> search(String query,
      [int page = 0, int perPage = 10]) async {
    try {
      var response = await dio.api.get('/auction/search/$query/$page/$perPage');
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Error searching auctions: $error $stackTrace');
      return [];
    }
  }

  Future<List<Auction>?> loadLatestAuctions() async {
    try {
      var response = await dio.api.get('/auction/latest');
      return (response.data as List)
          .map((auction) => Auction.fromJSON(auction))
          .toList();
    } catch (error, stackTrace) {
      print('Error getting latest auctions: $error $stackTrace');
      return null;
    }
  }

  Future<List<Auction>> getManySummary(List<String> auctionIds) async {
    try {
      var response = await dio.api.post('/auction/summary/many', data: {
        "auctionIds": auctionIds,
      });
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Error getting many summaries: $error, $stackTrace');
      return [];
    }
  }

  Future<List<AuctionMapData>> loadAuctionsForMap() async {
    try {
      var response = await dio.api.get(
        '/auction-map',
      );

      return List<AuctionMapData>.from(
        response.data.map(
          (el) => AuctionMapData.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Error getting auctions for map: $error $stackTrace');
      return [];
    }
  }

  Future<int> countFilter(
    List<String> categories,
    List<String> subCategories,
    List<String> locations,
    bool activeOnly, [
    bool includeMyAuctions = true,
    int? minPrice = 0,
    int? maxPrice = 0,
    String? usedCurrencyId,
  ]) async {
    try {
      var response = await dio.api.post('/auction/filter/count', data: {
        "categories": jsonEncode(categories),
        "subCategories": jsonEncode(subCategories),
        "locationIds": jsonEncode(locations),
        "activeOnly": activeOnly,
        "includeMyAuctions": includeMyAuctions,
        "minPrice": minPrice,
        "maxPrice": maxPrice,
        "usedCurrencyId": usedCurrencyId,
      });

      return int.parse(response.data['count'].toString());
    } catch (error) {
      print('Error counting filtered auctions: $error');
      return 0;
    }
  }

  Future<List<Auction>> loadActiveForAccount(
    String accountId, [
    int page = 0,
    int pageSize = 10,
    String query = '',
    AuctionsSortBy? sortBy = AuctionsSortBy.oldest,
  ]) async {
    try {
      String orderBy;
      String orderDirection;

      switch (sortBy) {
        case AuctionsSortBy.newest:
          orderBy = 'createdAt';
          orderDirection = 'desc';
        case AuctionsSortBy.oldest:
          orderBy = 'createdAt';
          orderDirection = 'asc';
        case AuctionsSortBy.priceDesc:
          orderBy = 'lastPrice';
          orderDirection = 'desc';
        case AuctionsSortBy.priceAsc:
          orderBy = 'lastPrice';
          orderDirection = 'asc';
        case null:
          orderBy = 'createdAt';
          orderDirection = 'desc';
      }

      var response =
          await dio.api.post('/auction/byAccount/active/$accountId', data: {
        "page": page,
        "perPage": pageSize,
        "query": query,
        "orderBy": orderBy,
        "orderDirection": orderDirection,
      });
      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error) {
      print('Error getting active auctions for account: $error');
      return [];
    }
  }

  Future<List<Auction>> loadFilteredAuctions(
    List<String> categories,
    List<String> subCategories,
    List<String> locations,
    bool activeOnly,
    int page,
    int pageSize, [
    String query = '',
    AuctionsSortBy? sortBy = AuctionsSortBy.oldest,
    bool includeMyAuctions = true,
    int? minPrice = 0,
    int? maxPrice = 0,
    bool? started = true,
    String? usedCurrencyId,
  ]) async {
    try {
      String orderBy;
      String orderDirection;

      switch (sortBy) {
        case AuctionsSortBy.newest:
          orderBy = 'createdAt';
          orderDirection = 'desc';
        case AuctionsSortBy.oldest:
          orderBy = 'createdAt';
          orderDirection = 'asc';
        case AuctionsSortBy.priceDesc:
          orderBy = 'lastPrice';
          orderDirection = 'desc';
        case AuctionsSortBy.priceAsc:
          orderBy = 'lastPrice';
          orderDirection = 'asc';
        case null:
          orderBy = 'createdAt';
          orderDirection = 'desc';
      }

      var response = await dio.api.post('/auction/filter/auctions', data: {
        "categories": jsonEncode(categories),
        "subCategories": jsonEncode(subCategories),
        "locationIds": jsonEncode(locations),
        "page": page,
        "perPage": pageSize,
        "activeOnly": activeOnly,
        "query": query,
        "orderBy": orderBy,
        "orderDirection": orderDirection,
        'includeMyAuctions': includeMyAuctions,
        'minPrice': minPrice,
        'maxPrice': maxPrice,
        'started': started,
        'usedCurrencyId': usedCurrencyId,
      });

      return List<Auction>.from(
        response.data.map(
          (el) => Auction.fromJSON(el),
        ),
      );
    } catch (error) {
      print('Error loading filtered auctions: $error');
      return [];
    }
  }

  Future<Auction?> update(CreateUpdateAuctionParams params) async {
    try {
      var data = await generateFormDataForCreateOrUpdate(params);
      var result = await dio.api.put('/auction/${params.id}', data: data);
      return Auction.fromJSON(result.data);
    } catch (error) {
      if (error is DioException) {
        var response = error.response?.data.toString() ?? '';

        if (response.contains('TOO_MANY')) {
          flashController.showMessageFlash(
              tr('create_auction.repository.too_many_assets'));
          return null;
        }

        if (response.contains('ASSET_TOO_BIG')) {
          flashController
              .showMessageFlash(tr('create_auction.repository.asset_too_big'));
          return null;
        }

        if (response.contains('ASSET_NOT_SUPPORTED')) {
          flashController.showMessageFlash(
              tr('create_auction.repository.asset_not_supported'));
          return null;
        }
      }

      flashController
          .showMessageFlash(tr('create_auction.repository.could_not_create'));
      return null;
    }
  }

  Future<Auction?> create(CreateUpdateAuctionParams params) async {
    try {
      var data = await generateFormDataForCreateOrUpdate(params);
      var result = await dio.api.post('/auction', data: data);
      return Auction.fromJSON(result.data);
    } catch (error) {
      if (error is DioException) {
        var response = error.response?.data.toString() ?? '';

        if (response.contains('TOO_MANY')) {
          flashController.showMessageFlash(
              tr('create_auction.repository.too_many_assets'));
          return null;
        }

        if (response.contains('ASSET_TOO_BIG')) {
          flashController
              .showMessageFlash(tr('create_auction.repository.asset_too_big'));
          return null;
        }

        if (response.contains('ASSET_NOT_SUPPORTED')) {
          flashController.showMessageFlash(
              tr('create_auction.repository.asset_not_supported'));
          return null;
        }
      }

      flashController
          .showMessageFlash(tr('create_auction.repository.could_not_create'));
      return null;
    }
  }

  Future<dio_lib.FormData> generateFormDataForCreateOrUpdate(
    CreateUpdateAuctionParams params,
  ) async {
    var galleryAssets = params.galleryAssets ?? [];
    var cameraAssets = params.cameraAssets ?? [];

    var files = galleryAssets.map((asset) async {
      var compressed =
          await imageCompressService.compressFileByPath(asset.path);
      return dio_lib.MultipartFile.fromFileSync(
        compressed.path,
        filename: asset.name,
      );
    }).toList();

    var cameraFiles = cameraAssets.map((asset) async {
      var compressed = await imageCompressService.compressFile(asset);
      return dio_lib.MultipartFile.fromFileSync(
        compressed.path,
        filename: asset.name,
      );
    }).toList();

    files.addAll(cameraFiles);

    var compressedFiles = await Future.wait(files);

    var data = dio_lib.FormData.fromMap({
      'latLng': jsonEncode(params.latLng),
      'location': params.location,
      'description': params.description ?? '',
      'hasCustomStartingPrice': params.hasCustomStartingPrice,
      'files': compressedFiles,
      'price': params.price,
      'startAt': params.startAt?.toIso8601String(),
      'youtubeLink': params.youtubeLink,
      'expiresAt': params.expiresAt?.toIso8601String(),
      'initialCurrencyId': params.initialCurrencyId,
      'title': params.title,
      // ignore: prefer_is_empty
      'assetsToKeep': params.assetsToKeep?.length != 0
          ? jsonEncode(params.assetsToKeep ?? [])
          : null,
      'mainCategoryId': params.mainCategoryId,
      'subCategoryId': params.subCategoryId,
      'condition': params.condition,
    });
    return data;
  }
}

class AuctionTranslationResult {
  final String? title;
  final String? description;

  AuctionTranslationResult({
    this.title,
    this.description,
  });

  static AuctionTranslationResult fromJSON(Map<String, dynamic> data) {
    return AuctionTranslationResult(
      title: data['title'],
      description: data['description'],
    );
  }
}
